File: common/src/main/java/wile/redstonepen/libmc/PlatformServices.java
package wile.redstonepen.libmc;

import java.util.ServiceLoader;

public class PlatformServices
{
  public static final INetworkingPlatform NETWORKING = load(INetworkingPlatform.class);
  public static final IPlatformHelper PLATFORM = load(IPlatformHelper.class);

  // Client-only — loaded lazily to avoid server-side class loading
  private static INetworkingClientPlatform networkingClient;
  private static IRenderingPlatform rendering;

  public static INetworkingClientPlatform getNetworkingClient()
  {
    if(networkingClient == null) networkingClient = load(INetworkingClientPlatform.class);
    return networkingClient;
  }

  public static IRenderingPlatform getRendering()
  {
    if(rendering == null) rendering = load(IRenderingPlatform.class);
    return rendering;
  }

  private static <T> T load(Class<T> clazz)
  {
    return ServiceLoader.load(clazz).findFirst()
      .orElseThrow(() -> new RuntimeException("No service implementation found for: " + clazz.getName()));
  }
}

File: common/src/main/kotlin/wile/redstonepen/platform/PlatformServices.kt
package wile.redstonepen.platform

import java.util.ServiceLoader

object PlatformServices {
    @JvmField val NETWORKING: INetworkingPlatform = load(INetworkingPlatform::class.java)

    @JvmField val PLATFORM: IPlatformHelper = load(IPlatformHelper::class.java)

    // Client-only — loaded lazily to avoid server-side class loading.
    // `by lazy {}` lambda type descriptors reference the client type at `<clinit>` time,
    // triggering NeoForge's dist-cleaner check before the lazy is ever accessed. Use
    // @Volatile backing fields instead to keep the reference out of class initialization.
    @Volatile private var networkingClient_: INetworkingClientPlatform? = null

    @Volatile private var rendering_: IRenderingPlatform? = null

    @JvmStatic
    fun getNetworkingClient(): INetworkingClientPlatform =
        networkingClient_
            ?: load(INetworkingClientPlatform::class.java).also { networkingClient_ = it }

    @JvmStatic
    fun getRendering(): IRenderingPlatform =
        rendering_ ?: load(IRenderingPlatform::class.java).also { rendering_ = it }

    private fun <T> load(clazz: Class<T>): T =
        ServiceLoader.load(clazz).findFirst().orElseThrow {
            RuntimeException("No service implementation found for: ${clazz.name}")
        }
}
